services/console-proxy/server/src/main/java/com/cloud/consoleproxy/vnc/NoVncClient.java

```diff
@@ -239,16 +273,349 @@ public byte[] encodePassword(byte[] challenge, String password) throws Exception
+        setWaitForNoVnc(true);
+        return verStr;
+    }
+
+    /**
+     * Once the protocol version has been decided, the server and client must agree on the type
+     * of security to be used on the connection.
+     *
+     * Reference: https://github.com/rfbproto/rfbproto/blob/master/rfbproto.rst#712security
+     */
+    public int handshakeSecurityType() {
+        while (isWaitForNoVnc()) {
+            s_logger.trace("Waiting for noVNC msg received");
+        }
+        s_logger.debug("Processing security types message");
+
+        int secType = RfbConstants.CONNECTION_FAILED;
+
+        List<Integer> secTypes = Arrays.asList(1, 2, 19, 261);
+
+        while (!socketConnection.checkIfBytesAreAvailableForReading(1)) {
+            s_logger.trace("Waiting for inStream to be ready");
+        }
+        int nServerSecTypes = socketConnection.readUnsignedInteger(8);
+        if (nServerSecTypes == 0) {
+            throw new CloudRuntimeException("No security types provided by the server");
+        }
+
+        Iterator<Integer> j;
+        for (int i = 0; i < nServerSecTypes; i++) {
+            int serverSecType = socketConnection.readUnsignedInteger(8);
+            s_logger.debug("Server offers security type " + serverSecType);
+
+            /*
+             * Use the first type sent by server which matches client's type.
+             * It means server's order specifies priority.
+             */
+            if (secType == RfbConstants.CONNECTION_FAILED) {
+                for (j = secTypes.iterator(); j.hasNext(); ) {
+                    int refType = (Integer) j.next();
+                    if (refType == serverSecType) {
+                        secType = refType;
+                        break;
+                    }
+                }
+            }
+        }
+        this.flushAfterReceivingNoVNCData = true;
+        setWaitForNoVnc(true);
+        return secType;
+    }
+
+    private final Object lock = new Object();
+    public void setWaitForNoVnc(boolean val) {
+        synchronized (lock) {
+            this.waitForNoVnc = val;
+        }
+    }
+
+    public boolean isWaitForNoVnc() {
+        synchronized (lock) {
+            return this.waitForNoVnc;
+        }
+    }
+
+    private boolean waitForNoVnc = false;
+
+    public void processSecurityResultMsg(int secType) {
+        s_logger.info("Processing security result message");
+        int result;
+
+        if (secType == RfbConstants.NO_AUTH) {
+            result = RfbConstants.VNC_AUTH_OK;
         } else {
-            os.write(b);
+            while (!socketConnection.checkIfBytesAreAvailableForReading(1)) {
+                s_logger.trace("Waiting for inStream");
+            }
+            result = socketConnection.readUnsignedInteger(32);
+        }
+
+        switch (result) {
+            case RfbConstants.VNC_AUTH_OK:
+                s_logger.info("Security completed");
```

Comment:
these message are a bit simple.
maybe can reuse the message in unused method `processSecurityResult`

```
            case RfbConstants.VNC_AUTH_TOO_MANY:
                s_logger.error("Connection to VNC server failed: too many wrong attempts.");
                throw new RuntimeException("Connection to VNC server failed: too many wrong attempts.");

            case RfbConstants.VNC_AUTH_FAILED:
                s_logger.error("Connection to VNC server failed: wrong password.");
                throw new RuntimeException("Connection to VNC server failed: wrong password.");

            default:
                s_logger.error("Connection to VNC server failed, reason code: " + authResult);
                throw new RuntimeException("Connection to VNC server failed, reason code: " + authResult);
        }
```